### PR TITLE
Update trackjet simulation

### PR DESCRIPTION
**Pull request by gkaratha**
#### PR description:
Dear all,

This PR contains all the improvements in the track jet code and the improved tagging of displaced jets (and their selection):
The work was presented in many GTT through the years but a summary is also presented in the last GTT meeting:
https://indico.cern.ch/event/1146596/ 

Several bug fixes are included and also an optimization of the code itself that makes the code 2 times faster than before. In the last slide there is a validation of the improved code.

Best regards,
George

### CODE REVIEW

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 1]
```diff
@@ -27,8 +29,10 @@
 #include "Geometry/CommonTopologies/interface/PixelGeomDetType.h"
 #include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
 
+#include "DataFormats/L1TCorrelator/interface/TkPrimaryVertex.h"
 #include "DataFormats/L1Trigger/interface/Vertex.h"
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:32]: Why add TkPrimaryVertex.h? It should be obsolete wrt Vertex.h

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 2]
```diff
@@ -61,73 +66,69 @@ class L1TrackJetProducer : public stream::EDProducer<> {
...
+  float trkZMax_;
+  float trkPtMax_;
+  float trkPtMin_;
+  float trkEtaMax_;
+  float nStubs4PromptChi2_;
+  float nStubs5PromptChi2_;
+  float nStubs4PromptBend_;
+  float nStubs5PromptBend_;
+  int trkNPSStubMin_;
+  int lowpTJetMinTrackMultiplicity_;
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:80]: Why not keep all these const?

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 3]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
       }
     }
+    // select tracks
     if (trk_nPS < trkNPSStubMin_)
-      continue;
+      continue; 
     if (!trackQualityCuts(trk_nstubs, trk_chi2dof, trk_bendchi2))
       continue;
-    if (std::abs(PVz - trk_z0) > dzPVTrk_)
+    if ( fabs(PVz-trkPtr->z0()) > dzPVTrk_ && dzPVTrk_>0 )
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: You should keep std::abs and not switch to fabs (CMSSW recommendation)

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 4]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
       continue;
-    if (std::abs(trk_z0) > trkZMax_)
+    if (fabs(trkPtr->z0()) > trkZMax_)
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Same here

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 5]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
       continue;
-    if (std::abs(trkPtr->momentum().eta()) > trkEtaMax_)
+    if ( fabs(trkPtr->momentum().eta())>trkEtaMax_)
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Same here

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 6]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
       continue;
     if (trk_pt < trkPtMin_)
       continue;
     L1TrkPtrs_.push_back(trkPtr);
-    zBinCount_.push_back(0);
 
-    if ((std::abs(trk_d0) > d0CutNStubs5_ && trk_nstubs >= 5 && d0CutNStubs5_ >= 0) ||
-        (trk_nstubs == 4 && std::abs(trk_d0) > d0CutNStubs4_ && d0CutNStubs4_ >= 0))
-      tdtrk_.push_back(1);  //displaced track
+    if ((fabs(trk_d0) > d0CutNStubs5_ && trk_nstubs >= 5 && d0CutNStubs5_>=0) || (trk_nstubs == 4 && fabs(trk_d0) > d0CutNStubs4_ && d0CutNStubs4_>=0))
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Same here

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 7]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
-              L2cluster[nclust].trackidx.push_back(trkidx1[itrk]);
-            }
-
-            if (used1 >= 0)
-              L1clusters[phibin + 1][used1].used = true;
-            if (used2 >= 0)
-              L1clusters[phibin + 1][used2].used = true;
-            nclust++;
+       } else{
+         if ( zmin > trkZ || zmin ==trkZ )
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Use ">="

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 8]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
+     L1clusters.push_back( L1_clustering(epbins[phibin],etaBins_,etaStep_) );
+   }
+
+   //second layer clustering in phi for all eta clusters  
+   L2clusters= L2_clustering(L1clusters, phiBins_, phiStep_, etaStep_);
+
+ // sum all cluster pTs in this zbin to find max
+   float sum_pt = 0;
+   for (unsigned int k = 0; k < L2clusters.size(); ++k) {
+     if (L2clusters[k].pTtot > 50 && L2clusters[k].numtracks < lowpTJetMinTrackMultiplicity_)
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Define the magic number 50 at the beginning of the code

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 9]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
         continue;
-      if (L2cluster[k].pTtot > minJetEtHighPt_ && L2cluster[k].numtracks < highpTJetMinTrackMultiplicity_)
+     if (L2clusters[k].pTtot > 100 && L2clusters[k].numtracks < highpTJetMinTrackMultiplicity_)
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Define the magic number 100

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 10]
```diff
@@ -142,509 +143,237 @@ L1TrackJetProducer::~L1TrackJetProducer() {}
...
+    int trk_nstub, float trk_chi2, float trk_bendchi2) {
   bool PassQuality = false;
-  if (trk_bendchi2 < trkBendChi2Max_ && trk_chi2 < trkChi2dofMax_ && trk_nstub >= 4 && !displaced_)
-    PassQuality = true;
-  if (displaced_ && trk_bendchi2 < nStubs4DisplacedBend_ && trk_chi2 < nStubs4DisplacedChi2_ && trk_nstub == 4)
-    PassQuality = true;
-  if (displaced_ && trk_bendchi2 < nStubs5DisplacedBend_ && trk_chi2 < nStubs5DisplacedChi2_ && trk_nstub > 4)
-    PassQuality = true;
+  if (!displaced_){
+     if ( trk_nstub == 4 && trk_bendchi2 < nStubs4PromptBend_ &&
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Define 4 and use it in the lines below too

File: L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py [hunk 11]
```diff
@@ -1,61 +1,77 @@
 import FWCore.ParameterSet.Config as cms
 from L1Trigger.VertexFinder.VertexProducer_cff import VertexProducer
 
-
+#prompt jet selection
 L1TrackJets = cms.EDProducer('L1TrackJetProducer',
 	L1TrackInputTag= cms.InputTag("TTTracksFromTrackletEmulation", "Level1TTTracks"),
-	L1PVertexCollection = cms.InputTag("VertexProducer", VertexProducer.l1VertexCollectionName.value()),
-	MaxDzTrackPV = cms.double( 0.5 ),
+#        L1TrackInputTag= cms.InputTag("TTTracksFromTrackletEmulation","Level1TTTracks","IN"),
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py]: Is it still needed (eg to revert back later)? If not please remove

File: L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py [hunk 12]
```diff
@@ -1,61 +1,77 @@
...
+        L1PVertexCollection = cms.InputTag("VertexProducer", VertexProducer.l1VertexCollectionName.value()),
+#        L1PVertexCollection = cms.InputTag("L1TkPrimaryVertex"),
```
Comment (cecilecaillol) [inline L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py]: Same here

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 1 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:32]: It was a relic left when i tried to compare f/w to simulation, and i forgot it. It has been removed now

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 2 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:80]: If I remember correctly, if we put const there we can change the value later. So probably is not the best option for on-the-fly selection

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 3 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: thank you, changed

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 7 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: thank you, changed

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 8 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Sure I will make it configurable. Thanks for catching it

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 10 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: I added the explanation: tracks with 4 stub are of poor resolution/quality therefore we put usually different cuts than those of high quality. This is why the variables have "Nstubs4" in the name

File: L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py (hunk 11 above)
Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/python/L1TrackJetProducer_cfi.py]: Yes that also i missed to remove from the period i ran the comparison. (F/W was compatible with old cmssw files, so it was the only way we could run.)

**Comment by gkaratha** [comment]
Dear @cecilecaillol ,

Thanks for the comments. I answered/changed everyting. I checked the code before and after the changes and the agreement is 100% as expected. The test was performed with my development branch 3.3.48.  I tried to run on the latest 3.4.23 but the *track* module (*not trackjet*) crashes with this error [*]. Possibly an update in the tracks themselves that happened between those two versions is to blame. Usually this is rather easy to solve by adding in my ntuplizer the correct source but it seems to have nothing to do with trackjets. 

Best,
George


[*] No "trackerDTC::SetupRcd" record found in the EventSetup.

**Comment by cecilecaillol** [comment]
Thanks @gkaratha! Have you run scram build code-checks and scram build code-format? If I merge this PR, is the track module going to crash for anyone using the code? Could you fix the code to run successfully on 3.4.23?

**Comment by gkaratha** [comment]
Hi @cecilecaillol ,

I did run those before making the PR, but just to be sure i rerun them today. Both times no errors were reported. This PR does not touches any of the track modules, is only trackjets so the state of the other code will remain as is. The crash i get is this [*]. I am guessing has something to do with the geometry of the sample or globaltag (or need to load some additional python module)? From the message though is evident that the module that crashes is the ttracks emulation and *not* the trackjet. Are you able to run the track emulation? If so, if you give the cfg file i can see what is the difference with mine. Thanks!



[*]----- Begin Fatal Exception 06-May-2022 12:26:12 CEST-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing global begin Run run: 1
   [1] Calling method for module L1FPGATrackProducer/'TTTracksFromExtendedTrackletEmulation'
Exception Message:
No "trackerDTC::SetupRcd" record found in the EventSetup.

**Comment by epalencia** [comment]
Hi @gkaratha, could you please try with l1t-phase2-v3.4.26 or newer?

**Comment by gkaratha** [comment]
Hi all,

So I i checked with 3.29 but the situation was the same. At the end I figured out that the missing ingredient was this process.load("L1Trigger.TrackerDTC.ProducerED_cff") and several other track related modules..  Also I changed the setup of tracker topology to use "ESGetToken", which was a warning msg printed when I used the latest  branch. 

Best,
george

**Comment by epalencia** [comment]
Tagged as **l1t-phase2-v3.4.30**.

**Pull request merged**